### PR TITLE
[Workflow] Add a line about the attribute type when persisted.

### PR DESCRIPTION
**Pull request by Guikingone**
By default, the attribute must be an array in order to allow PropertyAccess to update the state and content of this last one, if we need to persist it via Doctrine, the mapping must define an array.

### CODE REVIEW

**Comment by javiereguiluz** [comment]
@lyrixx could you please review this proposal? Thanks!

**Comment by lyrixx** [comment]
With a multiple marking store, it's indeed true.
But I'm not very confortable with this doc. The property name is `currentPlace`:
1. It's singular, but we says it should be an array
1. It's not the default name (`marking`), so it should be configured. Do we really want to overload the "getting started" documentation ?

**Comment by lyrixx** [comment]
About the issue:

* if a `MultipleStateMarkingStore` is used, the persisted state type **should be** an array. Or something that could represent many values (json_array, bit fields ...). Usually with doctrine, we use a `json` or `json_array` column.
* if a `SingleStateMarkingStore` is used, the persisted state type **could be** a `string`
* if something else is used, it's the responsibility of the creator of the "SomeThingElseMarkingStore" to know how to store the data

---

And juste to make thing really clear, the MarkingStore is not coupled to a Workflow / StateMachine.
A workflow, by nature will require a "MultipleMarkingStore". But if the workflow is "linear", there are not needs for a multiple marking store. But in 99.99% case, it will.
But as a subject managed by a StateMachine can be in only place, a SingleMarkingStore is enough.

**Comment by javiereguiluz** [comment]
Sadly this pull request has stalled. Should we reword it, create a different one or close it as "won't merge"? Thanks!

**Comment by Guikingone** [comment]
Yes, I need to have time to work on the reword, I'm gonna take some time this weekend in order to work on this PR.

**Comment by javiereguiluz** [comment]
Friendly ping to not forget about this. Thanks!

**Review by ndench — APPROVED** [review]
Looks good to me! Thanks for clarifying this.

File: workflow/usage.rst [hunk 1]
```diff
@@ -155,7 +155,16 @@ like this:
     The ``type`` (default value ``single_state``) and ``arguments`` (default value ``marking``)
     attributes of the ``marking_store`` option are optional. If omitted, their default values
     will be used.
+    
+.. note::
+
+    If a ``multiple_state`` is used, the persisted state type should be an ``array``, 
+    using Doctrine, you can use ``json`` or ``json_array``. 
+
+    If a ``single_state`` is used, the persisted state type could be a string
```
Comment (noniagriconomie) [inline workflow/usage.rst]: Or it could be also an integer, but you need to implement a MarkingStoreInterface to convert from string to int the place

Comment (Guikingone) [reply workflow/usage.rst]: Thanks for the feedback, I've added new lines about this case, can you review it and tell me if it's adapted? 🤔 

Thanks 😊

**Comment by javiereguiluz** [comment]
Should we close this pull request without merging it? First, the mentioned `json_array` type was deprecated in Doctrine. Second, the edge-case mentioned in `single_state` looks very edgy to me and I don't think we should mention it. Thoughts? Thanks!

**Comment by noniagriconomie** [comment]
@javiereguiluz 

maybe wait https://github.com/symfony/symfony/pull/29146 to be merged? and after document all cases?

but i really think a documentation (with example) is a must have for this part
actually lof of people using the workflow are using it with symfony, and with database storage

having a use case could be a nice one :)

**Comment by lyrixx** [comment]
No It should not be closed. This PR is good :+1:  Let's merge it (except for json_array)

File: workflow/usage.rst [hunk 2]
```diff
@@ -155,7 +155,19 @@ like this:
     The ``type`` (default value ``single_state``) and ``arguments`` (default value ``marking``)
     attributes of the ``marking_store`` option are optional. If omitted, their default values
     will be used.
+    
+.. note::
+
+    If a ``multiple_state`` is used, the persisted state type should be an ``array``, 
+    using Doctrine, you can use ``json`` or ``json_array``. 
+
+    If a ``single_state`` is used, the persisted state type could be a string,
```
Comment (javiereguiluz) [inline workflow/usage.rst]: "could be a string" or "should be a string" ... we say "should be an array" in the previous paragraph.

**Comment by Guikingone** [comment]
> No It should not be closed. This PR is good 👍 Let's merge it (except for json_array)

We could remove the `json_array` if it's a problem, not a big deal 😄

File: workflow/usage.rst [hunk 3]
```diff
@@ -155,7 +155,19 @@ like this:
...
+
+    If a ``multiple_state`` is used, the persisted state type should be an ``array``, 
+    using Doctrine, you can use ``json`` or ``json_array``. 
+
+    If a ``single_state`` is used, the persisted state type could be a string, 
+    if you need to store an int (due to DB limitations and/or project constraints),
+    be aware that you need to implement :class:`Symfony\\Component\\Workflow\\MarkingStore\\MarkingStoreInterface`
+    in order to transform the int back to something understandable by Workflow.
 
+    If something else is used, you're in charge of choosing the best storage format.
```
Comment (javiereguiluz) [inline workflow/usage.rst]: Does the previous tip related to MarkingStoreInterface apply in this case too? If yes, we must merge these two phrases.

File: workflow/usage.rst [hunk 4]
```diff
@@ -155,7 +155,19 @@ like this:
     The ``type`` (default value ``single_state``) and ``arguments`` (default value ``marking``)
     attributes of the ``marking_store`` option are optional. If omitted, their default values
     will be used.
+    
+.. note::
+
+    If a ``multiple_state`` is used, the persisted state type should be an ``array``,
```
Comment (javiereguiluz) [inline workflow/usage.rst]: This prhase needs some reword. It reads as "should be an array using Doctrine". Do you mean, a "PHP array" in the entity property and a "json" type in the `@Column` mapping?

File: workflow/usage.rst [hunk 5]
```diff
@@ -138,7 +138,7 @@ like this:
 
     class BlogPost
     {
-        // This property is used by the marking store
+        // This property is used by the marking store, by default, Symfony "mark" this attribute as an array.
```
Comment (javiereguiluz) [inline workflow/usage.rst]: I don't understand what this means --> _"Symfony mark this attribute as an array"_  Is this something that can only be understood if you use the Workflow component (which I don't) ... or is it supposed to be understood by all readers? Thanks!

Comment (lyrixx) [reply workflow/usage.rst]: May be this is better:
> By default symfony set the marking as an array. 
as or with, I don't know

File: workflow/usage.rst [hunk 6]
```diff
@@ -138,7 +138,7 @@ like this:
 
     class BlogPost
     {
-        // This property is used by the marking store
```
Comment (noniagriconomie) [inline workflow/usage.rst]: Is it the case with state machine as well?
State machine has only one place allowed at the same time

**Comment by javiereguiluz** [comment]
@Guikingone please, rebase this to 3.4 so we can merge it. Thank you.

**Comment by Guikingone** [comment]
@noniagriconomie @lyrixx 

Hi 👋 

Can you please check one last time this PR before the final rebase, this way, we can close it, thanks :)

**Comment by noniagriconomie** [comment]
@Guikingone there are some changes/new feature arround this storage part of the workflow
cf https://github.com/symfony/symfony/commit/59f20ad8fdd4d261fe752302918102adf60e0fb9
maybe this PR need to be closed, or adapted

File: workflow/usage.rst [hunk 7]
```diff
@@ -138,7 +138,7 @@ like this:
...
+        // This property is used by the marking store, by default, Symfony set the marking as an array.
```
Comment (lyrixx) [inline workflow/usage.rst]: To make things simple the current object is managed by: 
A state machine: It's a string
A workflow: it's an array

File: workflow/usage.rst (hunk 3 above)
Comment (lyrixx) [reply workflow/usage.rst]: @javiereguiluz yes it's apply to. So it should be merged

**Review by lyrixx — COMMENTED** [review]
Thanks for your PR. I added few comments.

File: workflow/usage.rst (hunk 7 above)
Comment (welcoMattic) [reply workflow/usage.rst]:
```suggestion
        // This property is used by the marking store. If the current object is managed by a state machine, this property is a string, else if it's managed by a workflow it's an array.
```

File: workflow/usage.rst (hunk 3 above)
Comment (lyrixx) [reply workflow/usage.rst]: @Guikingone could you merge theses phrase? Thanks.

After that, IMHO, we will be OK to merge it

Comment (Guikingone) [reply workflow/usage.rst]: Improved, to be checked 🙂

File: workflow/usage.rst [hunk 8]
```diff
@@ -138,24 +138,33 @@ like this:
...
     {
-        // This property is used by the marking store
         public $currentPlace;
         public $title;
         public $content;
     }
 
+.. tip::
+
+    The ``type`` (default value ``single_state``) and ``arguments`` (default value ``marking``)
```
Comment (noniagriconomie) [inline workflow/usage.rst]: For 4.3+ this need to be removed no?
As single/multiple state are deprecated/removed in favor of method

**Pull request closed without merging**